doc/reference/misc/data_structures.rst

```diff
@@ -0,0 +1,443 @@
+buffer with a specified byte count, where
+``RING_BUF_ITEM_DECLARE_SIZE()`` will declare and statically
+initialize a buffer with a given count of 32 bit words.
+``RING_BUF_ITEM_DECLARE_POW2()`` can be used to initialize an
+items-mode buffer with a memory region guaranteed to be a power of
+two, which enables various optimizations internal to the
+implementation.  No power-of-two initialization is available for
+bytes-mode ring buffers.
+
+"Bytes" data may be copied into the ring buffer using
+``ring_buf_put()``, passing a data pointer and byte count.  These
+bytes will be copied into the buffer in order, as many as will fit in
+the allocated buffer.  The total number of bytes copied (which may be
+fewer than provided) will be returned.  Likewise ``ring_buf_get()``
+will copy bytes out of teh ring buffer in the order that they were
+written, into a user-provided buffer, returning the number of bytes
+that were transferred.
+
+To avoid multiply-copied-data situations, a "claim" API exists for
+byte mode.  ``ring_buf_put_claim()`` takes a byte size value from the
+user and returns a pointer to memory internal to the ring buffer that
+can be used to receive those bytes, along with a size of the
+contiguous internal region (which may be smaller than requested).  The
+user can then copy data into that region at a later time without
+assembling all the bytes in a single region first.  When complete,
+``ring_buf_put_finish()`` can be used to signal the buffer that the
+transfer is complete, passing the number of bytes actually
+transferred.  At this point a new transfer can be initiated.
+Similarly, ``ring_buf_get_claim()`` returns a pointer to internal ring
+buffer data from which the user can read without making a verbatim
+copy, and ``ring_buf_get_finish()`` signals the buffer with how many
+bytes have been consumed and allows for a new transfer to begin.
+
+"Items" mode works similarly to bytes mode, except that all transfers
+are in units of 32 bit words and all memory is assumed to be aligned
+on 32 bit boundaries.  The write and read operations are
+``ring_buf_item_put()`` and ``ring_buf_item_get()``, and work
+otherwise identically to the bytes mode APIs.  There no "claim" API
+provided for items mode.  One important difference is that unlike
+``ring_buf_put()``, ``ring_buf_item_put()`` will not do a partial
+transfer; it will return an error in the case where the provided data
+does not fit in its entirety.
+
+The user can manage the capacity of a ring buffer without modifying it
+using the ``ring_buf_space_get()`` call (which returns a value of
+either bytes or items depending on how the ring buffer has been used),
+or by testing the ``ring_buf_is_empty()`` predicate.
+
+Finally, a ``ring_buf_reset()`` call exists to immediately empty a
+ring buffer, discarding the tracking of any bytes or items already
+written to the buffer.  It does not modify the memory contents of the
+buffer itself, however.
+
+Ring Buffer Internals
```

Comment:
You provided a diagram for the other data structures but not for this Ring Buffer...

Reply:
It's the least important (by far) thing documented here, and IMHO it's something where the physical representation (a blob of memory with two pointers into it) doesn't really lend itself to a diagram.  But I can try to cook something up.

Reply:
I'm OK without a diagram here then.